models/cryptopunks/ethereum/cryptopunks_ethereum_events.sql

```diff
@@ -0,0 +1,124 @@
+{{ config(
+        alias ='events',
+        materialized ='incremental',
+        file_format ='delta',
+        incremental_strategy='merge',
+        unique_key='unique_trade_id'
+        )
+}}
+
+WITH buys AS (
+  SELECT
+    `type`,
+    punkIndex,
+    CASE
+      WHEN buy_value = 0 THEN lag(bid_value) OVER (
+        PARTITION BY punkIndex
+        ORDER BY
+          evt_block_number ASC,
+          evt_index ASC
+      )
+      ELSE buy_value
+    END AS `enhanced_buy_value`,
+    `from`,
+    `bid_from`,
+    CASE
+      WHEN `to` = '0x0000000000000000000000000000000000000000' THEN lag(`bid_from`) OVER (
+        PARTITION BY punkIndex
+        ORDER BY
+          evt_block_number ASC,
+          evt_index ASC
+      )
+      ELSE `to`
+    END AS `enhanced_to`,
+    evt_block_time,
+    evt_block_number,
+    evt_index,
+    evt_tx_hash,
+    contract_address,
+    CASE WHEN (buy_value = 0 AND `to` = '0x0000000000000000000000000000000000000000') THEN "Offer Accepted" ELSE "Buy" END AS trade_category
+  FROM
+    (
+      SELECT
+        "PunkBought" AS `type`,
+        punkIndex,
+        value AS buy_value,
+        NULL AS bid_value,
+        toAddress as `to`,
+        fromAddress AS `from`,
+        NULL AS `bid_from`,
+        evt_block_number,
+        evt_index,
+        evt_block_time,
+        evt_tx_hash,
+        contract_address
+      FROM
+        {{ source('cryptopunks_ethereum','CryptoPunksMarket_evt_PunkBought') }}
+      UNION ALL
+        -- `from` is address that placed the bid
+      SELECT
+        "PunkBidEntered" AS `type`,
+        punkIndex,
+        NULL AS buy_value,
+        value AS bid_value,
+        NULL,
+        NULL AS `from`,
+        fromAddress AS `bid_from`,
+        evt_block_number,
+        evt_index,
+        evt_block_time,
+        evt_tx_hash,
+        contract_address
+      FROM
+        {{ source('cryptopunks_ethereum','CryptoPunksMarket_evt_PunkBidEntered') }}
```

Comment:
We might want to add incremental filters on decoded tables as well here: `evt_block_time >= date_trunc("day", now() - interval '1 week')`

Reply:
The outer function contains the lag() window function which requires all rows to be present to get accurate results. Not sure if other sql coding patterns are possible, but in the current setup incremental updates here seem not possible. Not a big issue here as the number of events is limited.